src/udev/udev-builtin-blkid.c

```diff
@@ -138,13 +138,49 @@ static int find_gpt_root(UdevEvent *event, blkid_probe pr) {
         assert(pr);
 
         /* Iterate through the partitions on this disk, and see if the UEFI ESP or XBOOTLDR partition we
-         * booted from is on it. If so, find the first root disk, and add a property indicating its partition
-         * UUID. */
+         * booted from is on it. If so, find the newest root partition, and add a property indicating its
+         * partition UUID. We also do this if we are dealing with a loopback block device whose "backing
+         * filename" field is set to the string "root". In the latter case we do not search for ESP or
+         * XBOOTLDR. */
+
+        const char *devtype;
+        r = sd_device_get_devtype(dev, &devtype);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to query devtype from device: %m");
+
+        if (!streq(devtype, "disk")) {
+                log_device_debug_errno(dev, r, "Skipping GPT root logic on partition block device.");
+                return 0;
+        }
+
+        sd_id128_t esp_or_xbootldr = SD_ID128_NULL;
+        bool need_esp_or_xbootldr;
+        r = efi_loader_get_device_part_uuid(&esp_or_xbootldr);
+        if (r < 0) {
+                if (r != -ENOENT && !ERRNO_IS_NOT_SUPPORTED(r))
+                        return log_debug_errno(r, "Unable to determine loader partition UUID: %m");
+
+                log_device_debug(dev, "No loader partition UUID EFI variable set, not using partition data for search for default root block device.");
+
+                /* NB: if an ESP/xbootldr field is set, we always use that. We do this in order to guarantee
+                 * systematic behaviour. */
+                if (!STRPTR_IN_SET(loop_backing_fname, "rootdisk", "rootdisk.raw")) {
+                        log_device_debug(dev, "Device is not a loopback block device with reference string 'root', not considering block device as default root block device.");
+                        return 0;
+                }
+
+                /* OK, we have now sufficiently identified this device as the right root "whole" device,
+                 * hence no need to bother with searching for ESP/XBOOTLDR */
+                need_esp_or_xbootldr = false;
+        } else
+                /* We now know the the ESP/xbootldr UUID, but we cannot be sure yet it's on this block
+                 * device, hence look for it among partitions now */
+                need_esp_or_xbootldr = true;
 
```

Comment:
Hmm, is this variable actually needed? Why not check sd_id128_id_null(esp_or_xbootldr)?

Reply:
no, it's not strictly needed, but i figured it's more readable this way, in particular with the if checks furthers down